examples/discovery.py

```diff
@@ -0,0 +1,82 @@
+import logging
+import aiohttp
+import aiohttp.web
+
+import asab
+import asab.web
+import asab.api
+import asab.zookeeper
+
+import os
+
+# `instance_id`` and `service_id` are identificators set as environemnt variables.
+# ASAB miroservices are meant to run in separate (Docker) containers.
+os.environ["INSTANCE_ID"] = "my_application_1"
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+asab.Config.add_defaults(
+	{
+		"web": {
+			"listen": "0.0.0.0 8089",
+		},
+
+		"zookeeper": {
+			"path": "example",
+			"servers": "localhost:2181"
+		},
+	}
+)
+
+
+class MyApplication(asab.Application):
+
+	async def initialize(self):
+		# Initialize web server
+		self.add_module(asab.web.Module)
+		websvc = self.get_service("asab.WebService")
+		self.WebContainer = asab.web.WebContainer(websvc, "web")
+
+		# Initialize zookeeper
+		self.add_module(asab.zookeeper.Module)
+		self.ZooKeeperService = self.get_service("asab.ZooKeeperService")
+		self.ZooKeeperContainer = asab.zookeeper.ZooKeeperContainer(self.ZooKeeperService, "zookeeper")
+
+		# Service Discovery is part of ApiService and needs its full functionality
+		self.ASABApiService = asab.api.ApiService(self)
+		self.ASABApiService.initialize_web(self.WebContainer)
+		self.ASABApiService.initialize_zookeeper(self.ZooKeeperContainer)
+
+		# Localize Service Discovery Service
+		self.DiscoveryService = self.get_service("asab.DiscoveryService")
+
+		self.WebContainer.WebApp.router.add_get('/locate', self.locate_self)
+
+	async def locate_self(self, request):
+		# This method seeks for this application in the ZooKeeper. Thus, it calls itself, being a tiny Oroboros.
+		# Try to run more than one ASAB Application with the same ZooKeeper configuration.
+		# You will get a tool to locate any service in the "cluster".
+
+		# Get config of the application:
+		config = None
+		# Pass the Application object to the DiscoverySession.
+		# The DiscoverySession is functional only with ApiService initialized.
+		async with asab.api.DiscoverySession(self) as session:
```

Comment:
I need the application object inside the session. And this is the easiest way...

Reply:
yes, that's cool

Reply:
I know.. I didn't know how to implement this one :( I'll read more about context managers